packages/playwright-core/src/server/injected/ariaSnapshot.ts

```diff
@@ -0,0 +1,240 @@
+      return;
+
+    const isElementVisible = injectedScript.utils.isElementVisible(element);
+    const hasVisibleChildren = element.checkVisibility({ opacityProperty: true, visibilityProperty: true, contentVisibilityAuto: true });
+
+    if (!hasVisibleChildren)
+      return;
+
+    if (isElementVisible) {
+      const childAriaNode = toAriaNode(element);
+      if (childAriaNode) {
+        ariaNode.children = ariaNode.children || [];
+        ariaNode.children.push(childAriaNode.ariaNode);
+      }
+      if (!childAriaNode?.isLeaf) {
+        for (let child = element.firstChild; child; child = child.nextSibling)
+          visit(childAriaNode?.ariaNode || ariaNode, child);
+      }
+    } else {
+      for (let child = element.firstChild; child; child = child.nextSibling)
+        visit(ariaNode, child);
+    }
+  };
+
+  injectedScript.utils.beginAriaCaches();
+  rootElement = rootElement || injectedScript.document.body;
+  const result = toAriaNode(rootElement);
+  const ariaRoot = result?.ariaNode || { role: '' };
+  try {
+    visit(ariaRoot, rootElement);
+  } finally {
+    injectedScript.utils.endAriaCaches();
+  }
+
+  normalizeStringChildren(ariaRoot);
+  return ariaRoot;
+}
+
+export function renderedAriaTree(injectedScript: InjectedScript, rootElement?: Element): string {
+  return renderAriaTree(injectedScript, generateAriaTree(injectedScript, rootElement));
+}
+
+function normalizeStringChildren(rootA11yNode: AriaNode) {
+  const flushChildren = (buffer: string[], normalizedChildren: (AriaNode | string)[]) => {
+    if (!buffer.length)
+      return;
+    const text = normalizeWhitespace(buffer.join(''));
+    if (text.trim())
+      normalizedChildren.push(text.trim());
+    buffer.length = 0;
+  };
+
+  const visit = (ariaNode: AriaNode) => {
+    const normalizedChildren: (AriaNode | string)[] = [];
+    const buffer: string[] = [];
+    for (const child of ariaNode.children || []) {
+      if (typeof child === 'string') {
+        buffer.push(child);
+      } else {
+        flushChildren(buffer, normalizedChildren);
+        visit(child);
+        normalizedChildren.push(child);
+      }
+    }
+    flushChildren(buffer, normalizedChildren);
+    ariaNode.children = normalizedChildren.length ? normalizedChildren : undefined;
+  };
+  visit(rootA11yNode);
+}
+
+const leafRoles = new Set([
+  'alert', 'blockquote', 'button', 'caption', 'checkbox', 'code', 'columnheader',
+  'definition', 'deletion', 'emphasis', 'generic', 'heading', 'img', 'insertion',
+  'link', 'menuitem', 'menuitemcheckbox', 'menuitemradio', 'meter', 'none', 'option',
+  'presentation', 'progressbar', 'radio', 'rowheader', 'scrollbar', 'searchbox', 'separator',
+  'slider', 'spinbutton', 'strong', 'subscript', 'superscript', 'switch', 'tab', 'term',
+  'textbox', 'time', 'tooltip'
+]);
+
+const normalizeWhitespace = (text: string) => text.replace(/[\s\n]+/g, ' ');
```

Comment:
We do have a different `normalizeWhitespace` function in `stringUtils` that handles zero-width whitespace and has a caching logic.